### ISSUE TITLE
Removing a container - UnsupportedSchemeException: https protocol is not supported

### ISSUE THREAD

**Issue by kiloalphabravotango** [bmuschko/gradle-docker-plugin#173]
Hi all

I'm getting this error when trying to remove a docker container (**DockerRemoveContainer** task) from a host on the local network.

My workflow is to use the **local** machine to build (via DockerBuildImage) and push (via DockerPushImage) the image to DockerHub, then on the **remote** machine, call DockerRemoveContainer then pull the recently created image from DockerHub so that I can run it.

The URL to the remote machine is one on the local network: http://192.168.202.31:2375
On both the local machine as well as the remote machine, the following variables were **not** set in the environment or the **/etc/sysconfig/docker** file:
- DOCKER_TLS_VERIFY
- DOCKER_CERT_PATH

I'm using the gradle-docker-plugin v2.6.6 with gradle v2.8.

The stacktrace is below. Strangely, it tries to remove the container, encounters problems, but then proceeds to try and pull the image:
```
Output: Removing container with ID 'kabtservice'.
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: Stopped existing container.
Output: :dockerPullImage
Output: Pulling repository 'kabtrepo/kabtservice'.
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] ERROR com.github.dockerjava.core.async.ResultCallbackTemplate - Error during callback
Output: org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
Output: at org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:99)
Output: at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:314)
Output: at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:365)
Output: at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:218)
Output: at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:194)
Output: at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:85)
Output: at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:108)
Output: at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
Output: at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:72)
Output: at com.github.dockerjava.jaxrs.connector.ApacheConnector.apply(ApacheConnector.java:441)
Output: at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:246)
Output: at org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
Output: at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:424)
Output: at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:679)
Output: at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:435)
Output: at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:338)
Output: at com.github.dockerjava.jaxrs.async.POSTCallbackNotifier.response(POSTCallbackNotifier.java:29)
Output: at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:45)
Output: at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:22)
Output: at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Output: at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Output: at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Output: at java.lang.Thread.run(Thread.java:745)
Output: :dockerPullImage FAILED
Output: 
Output: FAILURE: Build failed with an exception.
Output: 
Output: * What went wrong:
Output: Execution failed for task ':dockerPullImage'.
Output: > Could not pull image
```

Has anyone got any tips or seen this before? Thanks in advance.

Regards

KABT

**Comment by cdancy** [comment]
@kiloalphabravotango couple things first:

1.) Can you do these operations outside of the plugin successfully?
2.) Can you try with version 2.6.5 and see if that solves things? Shouldn't matter but folks have been noting issues with the latest release, that we are currently addressing, and it's possible those could have some effect here.
3.) Can you post your build.gradle file? It feels like this is an [incorrect setup for certPath](https://github.com/bmuschko/gradle-docker-plugin#extension-properties).

**Comment by kiloalphabravotango** [comment]
Hi @cdancy , thanks for getting back to me. 

**1.)**
Doing it manually seems fine (the specific container isnt running):
```
[root@java01 ~]# docker -H=tcp://192.168.202.31:2375 rm -f kabtservice
Error response from daemon: no such id: kabtservice
Error: failed to remove containers: [kabtservice]
[root@java01 ~]# 
```

**2.)**
I've tried v2.6.5 and unfortunately since then I've been gettin the following:
```
[pool-3-thread-1] ERROR com.github.dockerjava.core.async.ResultCallbackTemplate - Error during callback
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "aux" (class com.github.dockerjava.api.model.PushResponseItem), not marked as ignorable (9 known properties: , "stream", "error", "from", "progress", "time", "status", "errorDetail", "id", "progressDetail"])
 at [Source: com.github.dockerjava.jaxrs.util.WrappedResponseInputStream@4d99c05f; line: 154, column: 29] (through reference chain: com.github.dockerjava.api.model.PushResponseItem["aux"])
        at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:79)
        at com.fasterxml.jackson.databind.DeserializationContext.reportUnknownProperty(DeserializationContext.java:579)
        at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:672)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:906)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:328)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121)
        at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:2769)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1478)
        at com.github.dockerjava.core.async.JsonStreamProcessor.processResponseStream(JsonStreamProcessor.java:44)
        at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:62)
        at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:22)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**3.)** I thought this too but I also checked my certPath is not set.
My build.gradle is quite long, so I'm sending what I hope are relevant bits. Here is the docker closure, it sets the certPath with the dockerServerCertPath project property, but at this point its definitely null:
```
docker {
    // URL for Docker host's remote (Java) API
    url = project.hasProperty('dockerServerUrl') ? dockerServerUrl : null
    // Path to SSL certs needed to authenticate with Docker daemon if it's configured to require TLS verification
    // when communicating over HTTPS
    certPath = project.hasProperty('dockerServerCertPath') ? dockerServerCertPath : null

    // Credentials for authenticating with our private Docker registry when pushing or pulling Docker images
    registryCredentials {
        url = 'https://registry.hub.docker.com:443/v1'
        username = dockerHubUsername
        password = dockerHubPassword
        email = dockerHubEmail
    }
} 
```

Here is my DockerRemoveContainer task. It shows setting the certPath to null (just in case), however it is also null before that. I've made the DockerRemoveContainer task safe to execute if the container is not actually running:
```
task dockerRemoveContainer(type: DockerRemoveContainer) {
    force = true;
    targetContainerId { project.ext.serviceName }
}

task dockerSafelyRemoveContainer(dependsOn: 'dockerPushImage') {
    doFirst {
        try {
            docker.certPath = null;
            docker.url = dockerDeploymentServerUrl
            tasks.dockerRemoveContainer.execute()
        } catch (TaskExecutionException e) {
            if(!e.getCause() instanceof NotFoundException){
                throw e.getCause();
            }
            else{
                println "Stopped existing container."
            }
        }
    }
}
```

**Comment by cdancy** [comment]
So the failure is popping when you attempt to pull the image but the URL is https which appears to not be supported? For the 'dockerPullImage' task can you output your credentials in a 'doFirst {}' closure to check that the credentials are as they should be at that point in time? Looks like you are re-setting the extension at runtime and so maybe some values are getting used that are not expected?

**Comment by kiloalphabravotango** [comment]
The URL of the target machine where the existing container is to be stopped uses HTTP. 

The issue is occurring when I attempt to remove the container, then again when I try to pull a new image from DockerHub. Both of these commands are to be run on a remote machine (on the LAN).

As DockerHub uses HTTPs, do I need to set up certs in order to access **https://registry.hub.docker.com:443/v1**? I ruled this out because the error starts occuring before this, at the DockerRemoveContainer stage.

I've output my credentials in a doFirst{} for both the DockerRemoveContainer and DockerPullImage tasks:
```
task dockerRemoveContainer(type: DockerRemoveContainer) {
    force = true;
    targetContainerId { project.ext.serviceName }
}

task dockerSafelyRemoveContainer(dependsOn: 'dockerPushImage') {
    doFirst {
        try {
            println docker.registryCredentials.url;
            println docker.registryCredentials.username;
            println docker.registryCredentials.password;
            println docker.registryCredentials.email;
            docker.certPath = null;
            docker.url = dockerDeploymentServerUrl
            tasks.dockerRemoveContainer.execute()
        } catch (TaskExecutionException e) {
            if(!e.getCause() instanceof NotFoundException){
                throw e.getCause();
            }
            else{
                println "Stopped existing container."
            }
        }
    }
}

task dockerPullImage(type: DockerPullImage,dependsOn: ['dockerSafelyRemoveContainer'] ) {
    doFirst{
        println docker.registryCredentials.url;
        println docker.registryCredentials.username;
        println docker.registryCredentials.password;
        println docker.registryCredentials.email;
    }
    tag = project.ext.imageTagVersion
    repository= project.ext.imageTagName
}
```

However the issue still occurs (the real repository exists, I've just obfuscated the name here):
```
Output: :dockerBuildImage
Output: Building image using context '/var/tmp/build/kabtservice/build/docker'.
Output: Using tag 'kabtrepo/kabtservice:latest' for image.
Output: Created image with ID '9b93c2ed5fd8'.
Output: :dockerPushImage
Output: Pushing image with name 'kabtrepo/kabtservice'.
Output: :dockerSafelyRemoveContainer
Output: https://registry.hub.docker.com:443/v1
Output: kabtrepobuild
Output: kabtpassword
Output: build@kabtrepo.com
Output: Removing container with ID 'kabtservice'.
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [main] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: Stopped existing container.
Output: :dockerPullImage
Output: https://registry.hub.docker.com:443/v1
Output: kabtrepobuild
Output: kabtpassword
Output: build@kabtrepo.com
Output: Pulling repository 'kabtrepo/kabtservice'.
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request: https protocol is not supported
Output: [pool-3-thread-1] INFO org.apache.http.impl.execchain.RetryExec - Retrying request
Output: [pool-3-thread-1] ERROR com.github.dockerjava.core.async.ResultCallbackTemplate - Error during callback
Output: org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
Output: at org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:99)
Output: at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:314)
Output: at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:365)
Output: at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:218)
Output: at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:194)
Output: at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:85)
Output: at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:108)
Output: at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
Output: at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:72)
Output: at com.github.dockerjava.jaxrs.connector.ApacheConnector.apply(ApacheConnector.java:441)
Output: at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:246)
Output: at org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
Output: at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
Output: at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:424)
Output: at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:679)
Output: at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:435)
Output: at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:338)
Output: at com.github.dockerjava.jaxrs.async.POSTCallbackNotifier.response(POSTCallbackNotifier.java:29)
Output: at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:45)
Output: at com.github.dockerjava.jaxrs.async.AbstractCallbackNotifier.call(AbstractCallbackNotifier.java:22)
Output: at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Output: at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
Output: at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
Output: at java.lang.Thread.run(Thread.java:745)
Output: :dockerPullImage FAILED
```

**Comment by cdancy** [comment]
Do both of these task work on their own? Meaning if you just call the single task 'dockerPullImage', without it depending on any other task or you calling any other task, does it succeed?

**Comment by kiloalphabravotango** [comment]
Sorry for the delay, I was on vacation.

Running the steps manually shows that the remove and pull steps work correctly.
Thanks for pointing me in the right direction!

We were running this through a deployment system which was essentially calling "./gradlew deploy" with a few extra parameters. I took your advice and tried running the pull/remove tasks manually by connecting to the box. This worked, so I tried the exact same "deploy" goal that ran all the goals with their dependencies (just like the deployment system does). This also passed.

It appears that the issue I'm having is related to the user/environment that our deployment system is running. I'm not sure why as I printed the DOCKER_TLS_VERIFY and DOCKER_CERT_PATH variables from the same user which were not set. Also there were no entries for these in /etc/sysconfig/docker, so unfortunately it's not clear what is causing this problem only when the deployment system is running it, especially considering the fact that I ran the exact same command that it does.
